### PR TITLE
Notifications dropdown React integration

### PR DESCRIPTION
**Pull request by kntsoriano**
Closes https://github.com/exercism/v3-project-management/issues/127.

### CODE REVIEW

File: app/controllers/api/notifications_controller.rb [hunk 1]
```diff
@@ -9,7 +9,13 @@ def index
 
       serialized = SerializePaginatedCollection.(
         notifications,
-        SerializeUserNotifications
+        SerializeUserNotifications,
+        meta: {
```
Comment (kntsoriano) [inline app/controllers/api/notifications_controller.rb:13]: I added the ff to the API response:
- `links: all` in order to have a link to view all notifications
- `unread_count` in order to have the count of all unread notifications for the notification icon.

Is this okay?

File: app/channels/notifications_channel.rb [hunk 2]
```diff
@@ -1,9 +1,6 @@
 class NotificationsChannel < ApplicationCable::Channel
-  def self.broadcast_changed(user, count: user.notifications.unread.count)
-    NotificationsChannel.broadcast_to(user, {
-                                        type: "notifications.changed",
-                                        payload: { count: count }
-                                      })
+  def self.broadcast_changed(user)
+    NotificationsChannel.broadcast_to(user, { type: "notifications.changed" })
```
Comment (kntsoriano) [inline app/channels/notifications_channel.rb:3]: I've removed the count as I've changed the notifications dropdown to just refetch the API when a websockets message comes in.

File: app/controllers/api/notifications_controller.rb (hunk 1 above)
Comment (iHiD) [reply app/controllers/api/notifications_controller.rb:13]: Yep - this is fine :)

**Comment by iHiD** [comment]
@kntsoriano So to be clear on this, the notifications are pre-loaded into the dropdown, and are set on a page-load or websockets hook?

If that's the case, we also need to add a websockets push for when new badges are issues that this listens too as well.

If I've understood all that, it all seems sensible to me, so I'll happily merge :)

**Review by iHiD — APPROVED** [review]

**Comment by iHiD** [comment]
@kntsoriano I've thought of two problems with this approach. Neither block this PR, but, if you agree that they are problems, can you please open issues in v3-project-management assigned to you for fixing them before launch.

1. We rely very heavily on websockets here. That means that if a connection is dropped (say someone is going through a train tunnel) then they won't know they've got new notifications, even when they click the button and open the modal to check. This is an unexpected behaviour, as clicking the button on most sites pulls the data afresh. I think the easy option here is that when websockets reconnects (which I presume happens automatically?) then the subscription event should send the latest version of the notifications data in its return, which can then be updated.

2. As part of (1) and also in general with websockets, if the notifications dropdown is open when the notifications arrive, it is important that the HTML does not change for the user while they're looking at it. So I think we need a guard for this, that when a websocket comes in, if the dropdown is open, we hold the data and wait until the dropdown closes before applying it.

If I've misunderstood the code, and it pulls the data for the modal only on click, then (1) and may (2) might be redundant and you can ignore them. Let me know :)

**Comment by kntsoriano** [comment]
1. We can set a `staleTime` in React Query to ensure that it pulls notifications at a given interval.
2. Okay. This should be easy to do.

**Comment by kntsoriano** [comment]
@iHiD Notifications are set on page load and when websockets sends a message. It also refetches when the user switches windows / tabs (default react query behavior).

Yes. Pushing a new message when badges are unlocked should be done.

**Comment by SleeplessByte** [comment]
(react-query by default requeries on re-connect)

**Comment by iHiD** [comment]
> We can set a staleTime in React Query to ensure that it pulls notifications at a given interval.

Let's do this every minute please.

> react-query by default requeries on re-connect

Nice. Thanks @SleeplessByte. Is this related to a websockets connection, or just a general functionality of its awareness about connectivity? I don't know if the two are linked for react-query?

**Comment by SleeplessByte** [comment]
General connectivity connection, which you can "override" or amend to use custom triggers. So this can be solved.

**Comment by iHiD** [comment]
Interesting. I didn't realise it was so "alive".

**Pull request merged**